security-jwt/src/main/java/io/micronaut/security/token/jwt/validator/JwtTokenValidatorUtils.java

```diff
@@ -88,22 +89,21 @@ private JwtTokenValidatorUtils() {
         }
 
         final JWSAlgorithm algorithm = signedJWT.getHeader().getAlgorithm();
-        for (final SignatureConfiguration config : signatureConfigurations) {
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("Using signature configuration: {}", config.toString());
+
+        for (final SignatureConfiguration config : signatureConfigurations.stream()
+                .filter(c -> c.supports(algorithm))
+                .collect(Collectors.toList())) {
+            Optional<JWT> jwt = verifySignedJWTWithSignatureConfiguration(config, signedJWT);
+            if (jwt.isPresent()) {
+                return jwt;
             }
-            try {
-                if (config.verify(signedJWT)) {
-                    return Optional.of(signedJWT);
-                } else {
-                    if (LOG.isDebugEnabled()) {
-                        LOG.debug("JWT Signature verification failed: {}", signedJWT.getParsedString());
-                    }
-                }
-            } catch (final JOSEException e) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("Verification fails with signature configuration: {}, passing to the next one", config);
-                }
+        }
+        for (final SignatureConfiguration config : signatureConfigurations.stream()
```

Comment:
Does this make sense? Would a signature configuration that doesn't support the algorithm every verify the jwt?

Reply:
Originally, we only verified JWT signatures with signature configurations which appear to be compatible (`SignatureConfiguration::supports(_) >> true`).

 What was reported with https://github.com/micronaut-projects/micronaut-security/pull/95 is that there are some tokens which don't specify any algorithm in the header. Thus, we did not attempt to validate them against any signature configuration. For those JWTs the only way to ensure if there is a valid configuration is to try to verify it directly against the signature configuration. 

What this PR aims to achieve is to try to verify first with the signature configurations which support the specified algorithm and if no verification is possible try the rest.